sycl/include/sycl/ext/oneapi/group_sort.hpp

```diff
@@ -0,0 +1,159 @@
+#include <CL/sycl/detail/group_sort_util.hpp>
+#include <CL/sycl/detail/type_traits.hpp>
+#include <type_traits>
+
+#include "experimental/group_helpers_sorters.hpp"
+
+// define feature macro for group sorting algorithms
+#define SYCL_EXT_ONEAPI_GROUP_SORT 1
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace ext {
+namespace oneapi {
+namespace detail {
+
+// ---- traits
+template <typename T, typename = void> struct has_difference_type {};
+
+template <typename T>
+struct has_difference_type<T,
+#if __cplusplus < 201703
+                           cl::sycl::detail::void_type<
+#else
+                           std::void_t<
+#endif
+                               typename T::difference_type>> : std::true_type {
+};
+
+template <typename T> struct has_difference_type<T *> : std::true_type {};
+
+template <typename Sorter, typename Group, typename Val, typename = void>
+struct is_sorter_impl {
+  template <typename G>
+  using is_expected_return_type =
+      typename std::is_same<Val, decltype(std::declval<Sorter>()(
+                                     std::declval<G>(), std::declval<Val>()))>;
+
+  template <typename G = Group>
+  static decltype(
+      std::integral_constant<bool, is_expected_return_type<G>::value &&
+                                       sycl::is_group_v<G>>{})
+  test(int);
+
+  template <typename = Group> static std::false_type test(...);
+};
+
+template <typename Sorter, typename Group,
+          typename Ptr> // multi_ptr has difference_type and don't have other
+                        // iterator's fields
+struct is_sorter_impl<Sorter, Group, Ptr,
+#if __cplusplus < 201703
+                      cl::sycl::detail::void_type<
+#else
+                      std::void_t<
+#endif
+                          typename has_difference_type<Ptr>::type>> {
+  template <typename G = Group>
+  static decltype(std::declval<Sorter>()(std::declval<G>(), std::declval<Ptr>(),
+                                         std::declval<Ptr>()),
+                  sycl::detail::is_generic_group<G>{})
+  test(int);
+
+  template <typename = Group> static std::false_type test(...);
+};
+
+template <typename Sorter, typename Group, typename ValOrPtr>
+struct is_sorter : decltype(is_sorter_impl<Sorter, Group, ValOrPtr>::test(0)) {
+};
+} // namespace detail
+
+// ---- sort_over_group
+template <typename Group, typename T, typename Sorter>
+typename std::enable_if<detail::is_sorter<Sorter, Group, T>::value, T>::type
+sort_over_group(Group group, T value, Sorter sorter) {
+#ifdef __SYCL_DEVICE_ONLY__
+  return sorter(group, value);
+#else
+  (void)group;
+  throw runtime_error("Group algorithms are not supported on host device.",
+                      PI_INVALID_DEVICE);
+#endif
+}
+
+template <typename Group, typename T, typename Compare, std::size_t Extent>
+typename std::enable_if<!detail::is_sorter<Compare, Group, T>::value, T>::type
+sort_over_group(
+    cl::sycl::ext::oneapi::experimental::group_with_scratchpad<Group, Extent>
```

Comment:
I think it will still compile, while being much easier to read.

```suggestion
    group_with_scratchpad<Group, Extent>
```

Reply:
Are not fully qualified names preferred in the repo? If no, I can make it shorter. I'm just asking about conventions.

Reply:
Generally, we use fully qualified names when there can be discrepancies, like when SYCL defines the same type name as STL does. In this case `group_with_scratchpad` is unique to SYCL, so there's no need to duplicate namespaces.